net/Socket.hpp

```diff
@@ -1247,14 +1250,18 @@ class StreamSocket : public Socket,
     /// Emit 'onDisconnect' if it has not been done
     void ensureDisconnected()
     {
+        ASSERT_CORRECT_SOCKET_THREAD(this);
         if (!_doneDisconnect)
         {
-            ASSERT_CORRECT_SOCKET_THREAD(this);
-
             _doneDisconnect = true;
             if (_socketHandler)
                 _socketHandler->onDisconnect();
         }
+        if (!isShutdown())
```

Comment:
I suppose if we are in `ensureDisconnected()`, we must have `isShutdown()` independent of the conditionals above. Good.